tools/releng/go-prebuilt/go-build.sh

```diff
@@ -0,0 +1,199 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Copyright 2020 Joyent, Inc.
+#
+
+#
+# This script expects that you have downloaded and extracted a Go bootstrap
+# release from $GO_MANTA_URL and are running on a 'min_platform' SmartOS box.
+# It then expects to have a *.src.tar.gz release from $GO_SOURCE_URL and
+# have extracted it to a versioned directory at the top of your build dir.
+#
+# Then invoke using something like:
+#
+# $ ~/projects/triton.git/tools/releng/go-prebuilt/go-build.sh \
+#      -N 1.14 -b 1.13rc1 -d /ws/gobuild
+#
+#
+
+if [[ -n "$TRACE" ]]; then
+    export PS4='${BASH_SOURCE}:${LINENO}: '
+    set -o xtrace
+fi
+set -o errexit
+
+#
+# This URL is hardcoded in eng.git:tools/download_go. If it changes, be sure
+# to modify there too.
+#
+GO_MANTA_URL=https://us-east.manta.joyent.com/Joyent_Dev/public/releng/go/adhoc/
+
+#
+# The Golang source distribution download page
+#
+GO_SOURCE_URL=https://golang.org/dl/
+
+function usage {
+    echo ""
+    echo "Usage: go-build.sh [options]"
+    echo ""
+    echo "OPTIONS"
+    echo "  -b <version>        the bootstrap version to use"
+    echo "  -d <dir>            the top-level build directory"
+    echo "  -f                  force a build, ignoring min_platform and any existing output archive"
+    echo "  -N <version>        the new Golang version to build"
+    echo "  -C                  skip post-build cleanup and archive steps"
+    exit 2
+}
+
+function verify_min_platform {
+    MIN_PLATFORM=20181206T011455Z
+    CURRENT_PLATFORM=$(uname -v | sed -e 's/^joyent_//g')
+
+    if [[ "$CURRENT_PLATFORM" != "$MIN_PLATFORM" ]]; then
+        echo "The current platform $CURRENT_PLATFORM is not $MIN_PLATFORM"
+        echo "Golang builds for use in Triton/Manta should be done on the"
+        echo "oldest supported platform to ensure we're able to build with"
+        echo "it on any current build machine."
+        exit 1
+    fi
+}
+
+function build {
+    export GOROOT_BOOTSTRAP="$BUILD_DIR/$BOOTSTRAP_VER"
+    export GOROOT="$BUILD_DIR/$NEW_VER/proto"
+    export GOROOT_FINAL="/opt/go/$NEW_VER"
+    export GOOS='illumos'
+    export GOARCH='amd64'
+
+    cd "$BUILD_DIR/$NEW_VER/src"
+    echo "Building in dir: $(pwd)"
+    ./all.bash
+}
+
+function cleanup {
+    #
+    # Attempt to remove superfluous files to reduce the size
+    # of the Golang toolchain archive. This is based on code
+    # and comments in:
+    #     https://github.com/golang/build/blob/master/cmd/release/release.go
+    #
+    cd $BUILD_DIR/$NEW_VER
+
+    if [[ ! -f VERSION ]] || ! v=$(<VERSION) || [[ "$v" != "go${NEW_VER}" ]]; then
+        printf 'Are you in a go release directory?\n' >&2
+        exit 1
+    fi
+
+    # Remove the bootstrap compiler files.
+    rm -rf "pkg/bootstrap"
+
+    # Remove the obj build cache, which is _huge_.
+    rm -rf "pkg/obj"
+
+    # Remove the libraries used to build the toolchain commands.
+    rm -rf "pkg/illumos_amd64/cmd"
```

Comment:
This go-build.sh only needs to support newer Go versions that use the "illumos" GOOS, right?

Reply:
Yes - I'm assuming here that we always only want to build a go runtime for illumos. That's certainly the platform we'll be running on, and given that a go runtime will cross-compile, if we want to declare 'solaris' when building any Triton/Manta component at build-time, we can do eg.
```
GOOS=solaris go build
```
even with our illumos binary. For now, that's actually the default in Makefile.go_prebuilt.targ. I mentioned that in TOOLS-2439.